### PR TITLE
Rewrite StatusStream with rxcpp

### PR DESCRIPTION
**Pull request by lebdron**
Signed-off-by: Andrei Lebedev <lebdron@gmail.com>

### Description of the Change
Rewrite StatusStream method without manual synchronization with condition variables and mutexes using rxcpp

### Benefits
Separate status mapping and transport logic

### Possible Drawbacks 
Custom rxcpp timeout operator

### CODE REVIEW

File: irohad/torii/impl/command_service.cpp [hunk 1]
```diff
@@ -155,140 +168,104 @@ namespace torii {
...
 
-              if (isFinalStatus(resp_sub.tx_status())) {
-                response_writer.WriteLast(resp_sub, grpc::WriteOptions());
-                *finished = true;
-                cv->notify_one();
-              } else {
-                response_writer.Write(resp_sub);
-              }
+  rxcpp::observable<
+      std::shared_ptr<shared_model::interface::TransactionResponse>>
```
Comment (l4l) [inline irohad/torii/impl/command_service.cpp]: This type used multiple times, maybe create alias, e.g `using Response = std::shared_ptr<shared_model::interface::TransactionResponse>;`

File: irohad/torii/impl/command_service.cpp [hunk 2]
```diff
@@ -155,140 +168,104 @@ namespace torii {
...
+        .lift<iroha::protocol::ToriiResponse>(
+            makeTimeout<iroha::protocol::ToriiResponse>(
+                [&](const auto &response) {
+                  return response.tx_status()
+                          == iroha::protocol::TxStatus::NOT_RECEIVED
+                      ? start_tx_processing_duration_
+                      : 2 * proposal_delay_;
+                },
+                current_thread))
+        .take_while([=](const auto &) { return not context->IsCancelled(); })
```
Comment (l4l) [inline irohad/torii/impl/command_service.cpp]: Might be useful to add logging here as well

File: irohad/torii/impl/timeout.hpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Comment (l4l) [inline irohad/torii/impl/timeout.hpp]: I guess there should be one more license

File: irohad/torii/impl/timeout.hpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+   * @tparam Coordination the type of the scheduler
+   */
+  template <class T, class Selector, class Coordination>
+  struct timeout {
+    typedef rxcpp::util::decay_t<T> source_value_type;
+    typedef rxcpp::util::decay_t<Coordination> coordination_type;
+    typedef typename coordination_type::coordinator_type coordinator_type;
+    typedef rxcpp::util::decay_t<Selector> select_type;
+    typedef decltype(
+        (*(select_type *)nullptr)(*(source_value_type *)nullptr)) value_type;
```
Comment (l4l) [inline irohad/torii/impl/timeout.hpp]: Is it used?
Also declval might be applicable

File: irohad/torii/impl/timeout.hpp [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+          localState->worker.schedule(produce_time,
+                                      produce_timeout(new_id, localState));
+        };
+        auto selectedWork =
+            on_exception([&]() { return localState->coordinator.act(work); },
+                         localState->dest);
+        if (selectedWork.empty()) {
+          return;
+        }
+        localState->worker.schedule(selectedWork.get());
```
Comment (l4l) [inline irohad/torii/impl/timeout.hpp]: Maybe invert `if` above and move this line up?

File: irohad/torii/impl/command_service.cpp [hunk 6]
```diff
@@ -44,9 +46,15 @@ namespace torii {
         proposal_delay_(proposal_delay),
         start_tx_processing_duration_(1s),
         cache_(std::make_shared<CacheType>()),
+        // merge with mutex, since notifications can be made from different
+        // threads
+        responses_(tx_processor_->transactionNotifier().merge(
```
Comment (muratovv) [inline irohad/torii/impl/command_service.cpp]: Pls, add TODO about event bus rework.

File: irohad/torii/impl/command_service.cpp [hunk 7]
```diff
@@ -155,140 +168,104 @@ namespace torii {
...
+              return shared_model::proto::TransactionStatusBuilder()
+                  .txHash(hash)
+                  .notReceived()
+                  .build()
+                  .getTransport();
+            }())));
+    return responses_.start_with(initial_status)
+        .filter(
+            [&](auto response) { return response->transactionHash() == hash; })
+        .lift<std::shared_ptr<shared_model::interface::TransactionResponse>>(
```
Comment (muratovv) [inline irohad/torii/impl/command_service.cpp]: I think, better to add some comments about what happened here.

File: irohad/torii/impl/command_service.cpp [hunk 8]
```diff
@@ -155,140 +168,104 @@ namespace torii {
...
+                     shared_model::proto::TransactionResponse>(response)
+              ->getTransport();
+        })
+        .lift<iroha::protocol::ToriiResponse>(
+            makeTimeout<iroha::protocol::ToriiResponse>(
+                [&](const auto &response) {
+                  return response.tx_status()
+                          == iroha::protocol::TxStatus::NOT_RECEIVED
+                      ? start_tx_processing_duration_
+                      : 2 * proposal_delay_;
```
Comment (muratovv) [inline irohad/torii/impl/command_service.cpp]: maybe move out of here configuration logic?

File: irohad/torii/impl/command_service.cpp [hunk 9]
```diff
@@ -155,140 +168,104 @@ namespace torii {
...
+                },
+                current_thread))
+        .take_while([=](const auto &) { return not context->IsCancelled(); })
+        .subscribe(
+            subscription,
+            [&](iroha::protocol::ToriiResponse response) {
+              log_->debug("writing status");
```
Comment (muratovv) [inline irohad/torii/impl/command_service.cpp]: Move the debug after `Write` method.

File: irohad/torii/impl/command_service.cpp [hunk 10]
```diff
@@ -155,140 +168,104 @@ namespace torii {
...
-                      request_hash->hex(),
-                      resp->tx_status());
-        }
-      }
-    } else {
-      log_->debug("StatusStream request processed successfully, hash: {}",
-                  request_hash->hex());
+    // run loop while subscription is active or there are pending events in the
+    // queue
+    while (subscription.is_subscribed() or not rl.empty()) {
```
Comment (muratovv) [inline irohad/torii/impl/command_service.cpp]: better to wrap this function with an inline method for the improving readability.

File: irohad/torii/impl/timeout.hpp [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+#ifndef IROHA_TORII_IMPL_TIMEOUT_HPP
+#define IROHA_TORII_IMPL_TIMEOUT_HPP
+
+#include <rxcpp/operators/rx-timeout.hpp>
+
+namespace torii {
+
+  /**
+   * Return an observable that terminates with timeout_error if a particular
```
Comment (muratovv) [inline irohad/torii/impl/timeout.hpp]: Maybe better to write here about "This class is copypaste of ..." for reducing misleading usages of code as an example.

File: irohad/torii/impl/timeout.hpp [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_TORII_IMPL_TIMEOUT_HPP
+#define IROHA_TORII_IMPL_TIMEOUT_HPP
+
+#include <rxcpp/operators/rx-timeout.hpp>
+
+namespace torii {
```
Comment (muratovv) [inline irohad/torii/impl/timeout.hpp]: Maybe move this file to lib for sharing usage?

File: irohad/torii/impl/command_service.cpp [hunk 13]
```diff
@@ -234,38 +236,53 @@ namespace torii {
...
+                      ? initial_timeout_
+                      : nonfinal_timeout_;
                 },
                 current_thread))
-        .take_while([=](const auto &) { return not context->IsCancelled(); })
+        // complete the observable if client is disconnected
+        .take_while([=](const auto &) {
+          auto is_cancelled = context->IsCancelled();
+          if (is_cancelled) {
+            log_->debug("client unsubscribed");
```
Comment (muratovv) [inline irohad/torii/impl/command_service.cpp]: Maybe specify here some identify information about a client?

File: irohad/torii/impl/command_service.cpp [hunk 14]
```diff
@@ -234,38 +236,53 @@ namespace torii {
...
+          }
+          return not is_cancelled;
+        })
         .subscribe(
             subscription,
             [&](iroha::protocol::ToriiResponse response) {
-              log_->debug("writing status");
-              response_writer->Write(response);
+              if (response_writer->Write(response)) {
+                log_->debug("status written");
```
Comment (muratovv) [inline irohad/torii/impl/command_service.cpp]: same as above

**Pull request merged**